### PR TITLE
frontend: add a dot-com API endpoint caching src versions by branch

### PR DESCRIPTION
**Pull request by LawnGnome**
This PR adds a simple version cache to the unauthenticated external REST API that reports the most recent version of src-cli for the given branch. This is the backend that will address #39468, but #40828 also has to be merged for this to be fully fixed.

The intention is that the sourcegraph.com deployment will enable the cache, and no other deployments will ever configure it.

The minimal configuration that will be required on sourcegraph.com will be:

```json
{
  "dotcom": {
    "srcCliVersionCache": {
      "enabled": true,
      "github": {
        "token": "TOKEN",
        "webhookSecret": "WEBHOOK_SECRET"
      }
    }
  }
}
```

This will need to be combined with a GitHub webhook on https://github.com/sourcegraph/src-cli that points to https://sourcegraph.com/.api/src-cli/versions/webhook, configured with the same `WEBHOOK_SECRET` and with the `release` event type enabled.

### Why not use the `SOURCEGRAPHDOTCOM_MODE` environment variable?

This PR adds a small configuration section to the `dotcom` site configuration object that enables and configures the cache, but does _not_ use the `SOURCEGRAPHDOTCOM_MODE` environment variable.

This is both to simplify testing and reduce the matrix of possible configuration options. With this implementation, there are only two possible states: configuration present and enabled, in which case this API will respond, and literally anything else, in which case this API will 404 like any other unknown endpoint.

### Why is the handler always installed?

Originally, I had the handler only conditionally hooked up if the configuration was enabled. However, it's considerably simpler if it's always present, since it means we can handle all the enable/disable behaviour in a single configuration watcher, and from a user perspective the disabled handler is indistinguishable from the route not existing in the router.

## Test plan

This has (IMHO) good test coverage, and I've given it a pretty good hammering in the last day or so locally as part of developing this and #40828.

### CODE REVIEW

File: internal/extsvc/github/v4.go [hunk 1]
```diff
@@ -700,6 +700,93 @@ func (c *V4Client) RecentCommitters(ctx context.Context, params *RecentCommitter
...
+	if err != nil {
+		var e graphqlErrors
+		if errors.As(err, &e) {
+			for _, err2 := range e {
+				if err2.Type == graphqlErrTypeNotFound {
+					c.log.Warn("GitHub repository not found", graphQLErrorField(err2))
+					continue
+				}
+				return nil, err
+			}
```
Comment (eseliger) [inline internal/extsvc/github/v4.go:781]: it's late here, but isn't there a path where none of the errs are the not found error where we just swallow the error and continue?

File: cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  "3.40": "3.40.11",
+  "3.41": "3.41.1",
+  "3.42": "3.42.3",
+  "3.43": "3.43.1",
+  "3.5": "3.5.0",
+  "3.6": "3.6.1",
+  "3.7": "3.7.1",
+  "3.8": "3.8.0",
+  "3.9": "3.9.0",
+  "4.30": "4.30.9"
```
Comment (eseliger) [inline cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch:50]: heck, why are these here?

File: cmd/frontend/internal/httpapi/releasecache/http.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+		// already stopped any updater that was running.
+		handler.enabled = config.enabled
+		if !handler.enabled {
+			return
+		}
+
+		// Otherwise, let's build a new release cache and start a fresh updater.
+		rc := config.NewReleaseCache(logger)
+		handler.updater = goroutine.NewPeriodicGoroutine(ctx, config.interval, rc)
+		go goroutine.MonitorBackgroundRoutines(ctx, handler.updater)
```
Comment (eseliger) [inline cmd/frontend/internal/httpapi/releasecache/http.go:75]: does this free up all resources when all the passed handlers are stopped or will it keep a dangling go routine?

File: cmd/frontend/internal/httpapi/releasecache/http.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func (h *handler) handleWebhook(w http.ResponseWriter, r *http.Request) {
+	h.doHandleWebhook(w, r, gh.ValidateSignature)
+}
+
+type signatureValidator func(signature string, payload []byte, secret []byte) error
+
+func (h *handler) doHandleWebhook(w http.ResponseWriter, r *http.Request, signatureValidator signatureValidator) {
+	payload, err := io.ReadAll(r.Body)
```
Comment (eseliger) [inline cmd/frontend/internal/httpapi/releasecache/http.go:142]: do we need to close r.Body?

File: schema/site.schema.json [hunk 5]
```diff
@@ -1435,6 +1435,60 @@
...
           "group": "Sourcegraph.com"
+        },
+        "srcCliVersionCache": {
+          "description": "Configuration related to the src-cli version cache. This should only be used on sourcegraph.com.",
+          "type": "object",
+          "required": ["enabled", "github"],
+          "group": "Sourcegraph.com",
+          "properties": {
+            "enabled": {
+              "description": "Enables the src-cli current version API endpoint.",
```
Comment (eseliger) [inline schema/site.schema.json:1446]: this just toggles the cache pass-through endpoint, not the endpoint to get the current version, correct?

**Review by eseliger — APPROVED** [review]
Wow, so much stuff you implemented here, even webhooks to make it close to real time! Great work, also on all the testing!

File: cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch (hunk 2 above)
Comment (LawnGnome) [reply cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch:50]: This was a mistaken release a few months ago. I have now removed it.

File: cmd/frontend/internal/httpapi/releasecache/http.go (hunk 3 above)
Comment (LawnGnome) [reply cmd/frontend/internal/httpapi/releasecache/http.go:75]: They [should get cleaned up](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@a638173f6582d5531a2dffccafaade5de21e4bea/-/blob/internal/goroutine/background.go?L59-65), unless I'm missing something. (The captured `rc` will be garbage collected once the handler is destroyed and `handler.rc` no longer holds a reference to it.)

File: cmd/frontend/internal/httpapi/releasecache/http.go (hunk 4 above)
Comment (LawnGnome) [reply cmd/frontend/internal/httpapi/releasecache/http.go:142]: Probably. (I'm never 100% clear on the rules around that in Go — it still feels like a total misfeature to me that the client isn't responsible for this.)

File: schema/site.schema.json (hunk 5 above)
Comment (LawnGnome) [reply schema/site.schema.json:1446]: Correct!

File: internal/extsvc/github/v4.go (hunk 1 above)
Comment (LawnGnome) [reply internal/extsvc/github/v4.go:781]: It does seem like it. (I copy pasted this from `RecentCommitters`, honestly, who copy pasted this from `GetReposByNameWithOwner`.)

I'll fix those up along with `Releases`.

Comment (LawnGnome) [reply internal/extsvc/github/v4.go:781]: It turns out `GetReposByNameWithOwner` actually relies on this. 🙈 

The others are definitely wrong, though, so I'm fixing those.

Comment (eseliger) [reply internal/extsvc/github/v4.go:780]: can you also add a comment to `GetReposByNameWithOwner` to stop the blindfolded copy-pasting? 😬

Comment (LawnGnome) [reply internal/extsvc/github/v4.go:780]: Sure thing.

File: cmd/frontend/internal/httpapi/releasecache/cache.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	}
+}
+
+type branchNotFoundError string
+
+func (e branchNotFoundError) Error() string {
+	return "branch not found: " + string(e)
+}
+
+func (e branchNotFoundError) NotFound() bool { return true }
```
Comment (Piszmog) [inline cmd/frontend/internal/httpapi/releasecache/cache.go:137]: Does not look like this is used anywhere

File: cmd/frontend/internal/httpapi/releasecache/cache.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+		// release order, we don't have to do any version comparisons: we can
+		// simply use the first release on the branch only and ignore the rest.
+		branch := fmt.Sprintf("%d.%d", version.Major, version.Minor)
+		if _, found := branches[branch]; !found {
+			branches[branch] = release.TagName
+		}
+	}
+}
+
+type branchNotFoundError string
```
Comment (Piszmog) [inline cmd/frontend/internal/httpapi/releasecache/cache.go:131]: I was wondering whether this error actually buys anything. I see it is used for tests (`assert.ErrorAs`). But I believe those could be changed to `assert.ErrorIs(t, err, errors.New("branch not found: " + branch))`

**Review by Piszmog — APPROVED** [review]

File: cmd/frontend/internal/httpapi/releasecache/cache.go (hunk 6 above)
Comment (LawnGnome) [reply cmd/frontend/internal/httpapi/releasecache/cache.go:137]: It gets used indirectly here: https://github.com/sourcegraph/sourcegraph/pull/40827/files#diff-371c45bce51b6559b26cee919ae3e217dcf6d76e0fb348fd346f5a4aa6980b1eR115

File: cmd/frontend/internal/httpapi/releasecache/cache.go (hunk 7 above)
Comment (LawnGnome) [reply cmd/frontend/internal/httpapi/releasecache/cache.go:131]: (see above)

File: cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch (hunk 2 above)
Comment (courier-new) [reply cmd/frontend/internal/httpapi/releasecache/testdata/golden/TestReleaseCache_Fetch]: Mistaken release DEFINITELY wasn't me or anything. 👀
